### DISCUSSION TITLE
Dudas sobre before

### DISCUSSION THREAD

**Discussion by heinHertz** [dds-jv/foro#162] [Consultas de UI / REST] [answered]
Hola necesito una mano acá,  quiero hacer el logeo o control del usuario desde el before de sparks  donde están las rutas y methods , el tema es que no le agarré la mano o me está costando solucionar ese tema. quiero que una vez en el before colocar algun tipo de filtro o control que me desvie a "/login"  en caso de que el usuario actual no este logueado .  lo que busco en realidad es algun control antes de acceder a cualquier ruta de confirmacion de login.
hasta el momento para loguear me funciona bien mi ruta "login" y accedo a la pagina ( al sistema ) después de loguear, pero si pongo ese método  para autenticar  en el before me entra en un bucle infinito que no pude resolver todavía. 
acá el pseudocódigo ( no esta asi en mi código pero da la idea) de login

```
method  login(Request request, Response response )

    String username = request.queryParams("usuario")
    String password = request.queryParams("contrasenia")

    if (    RepositorioUsuarios.estaElsuario(usarname, password ) )
       RepositorioUsuarios.usuariosLogueados.add(usuario)
       request.session().attribute("idUsuario", usuario.getId_usuario())
       response.redirect("/operaciones")
    else
		response.redirect("/login")
```

el tema es que que no se como usar el before o que recomendación debería tener para hacer
si hago algo así 

```
before("/*", (request, response) ->
    Long idUsuario = request.session().attribute("idUsuario")

    if ( idUsuario == null ) 	<-----  o algo que confirme que no existe o que no esta  logueado		
	redirect("/login")			
    }

```
me redirige a un bucle infinito , no se que hacer acá abajo el código de logeo, muy simple por cierto.  Todos en LoginController
```
public static boolean estaLogueado(Request request, Response response) {
		Optional<Usuario> usuario = getUsuarioLogueado(request);

		return usuario.isPresent();
	}

	public static Optional<Usuario> getUsuarioLogueado(Request request) {
		Long idUsuario = request.session().attribute("idUsuario");

		Optional<Usuario> usuario = Optional.empty();

		if (idUsuario != null) {

			usuario = RepositorioUsuarios
					.getInstance()
					.usuariosLogueados()
					.stream()
					.filter(u -> u.getId_usuario().equals(idUsuario)).findFirst();
		}

		return usuario;

	}

	public static void asegurarseUsuarioEstaLoggeado(Request request, Response response) {

		if (!LoginController.estaLogueado(request, response)) {
			response.redirect("/login");
		}
```

si saco todos los "redirect( "/login" ) del LoginController también entra en bucle infinito, así que no se que puede ser

muchas gracias

**Comment by gastonprieto** [comment]
Buenas @heinHertz !

Creo que tu problema puntual es que lo valida en todas las rutas, entonces lo que deberías hacer es excluir a la ruta de '/login' de la autenticación. A la request le podes pedir cual es el path para ignorar la validación en el caso que no corresponda.

**Comment by gastonprieto** [comment]
Otros dos comentarios que no tienen que ver con tu duda puntal son:

- `RepositorioUsuarios.usuariosLogueados.add(usuario)` no entiendo` porque lo esta agregando en el repositorio de usuario, si en realidad la session es la que te esta marcando si esta logueado o no.
- Trata de ponerle "```" al código para que te lo formatee (yo te editee el issue y te lo puse)
- Trata de no adjuntar imagenes, porque luego si alguien quiere buscar algo parecido no lo va a encontrar.

**Comment by gastonprieto** [comment]
Buenas @heinHertz te sirvio? te quedo alguna duda sobre esto?

**Comment by heinHertz** [comment]
todavía no pude, cuando pueda subo

**Comment by heinHertz** [comment, marked as answer]
listo encontré esta solucion:

```Java
before("/*", (request, response) -> {
		   
	if(!request.pathInfo().equals("/login"))
	LoginController.asegurarseUsuarioEstaLoggeado(request, response);
		
});
```
Gracias.
